LocalStorage api

msau42 opened the pull request:
**What this PR does / why we need it**:
API changes to support persistent local volumes, as described [here](https://github.com/kubernetes/community/pull/306)

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
Part of #43640

**Special notes for your reviewer**:
There were a few items I was concerned about.  Will add review comments in those places.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
NONE

Note will be added in subsequent PR with the volume plugin changes

msau42 committed:
API changes for persistent local volumes.

Includes:
- A new volume type, LocalVolumeSource.  This only supports
file-based local volumes for now.
- New alpha annotation in PV: NodeAffinity
- Validation + tests for specifying LocalVolumeSource and PV
NodeAffinity
- Alpha feature gate

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44640)
<!-- Reviewable:end -->

msau42 commented on pkg/api/validation/validation.go:
```diff
@@ -1237,6 +1261,23 @@ func ValidatePersistentVolume(pv *api.PersistentVolume) field.ErrorList {
+			numVolumes++
+			allErrs = append(allErrs, validateLocalStorageVolumeSource(pv.Spec.LocalStorage, specPath.Child("localStorage"))...)
+			// StorageClass is required
+			if len(pv.Spec.StorageClassName) == 0 {
```
What I actually want to verify is that the StorageClass exists and also has the topology-key set.  I'm not sure if we can do that at this point though.

msau42 commented:
cc @vishh @kubernetes/sig-storage-api-reviews @kubernetes/sig-storage-pr-reviews

msau42 commented on pkg/api/types.go:
```diff
@@ -302,6 +302,11 @@ type VolumeSource struct {
 	// +optional
 	ScaleIO *ScaleIOVolumeSource
+	// LocalStorage represents a local volume accessible only from one node
+	// This volume type cannot be used as a direct volume source and is only required here
```
Just checked with Saad.  ConstructVolumeSpec can actually support PersistentVolumeSource too, so then I can remove this from the Volume Source.  :-D

jsafrane commented on pkg/api/validation/validation.go:
```diff
@@ -1237,6 +1261,23 @@ func ValidatePersistentVolume(pv *api.PersistentVolume) field.ErrorList {
+				allErrs = append(allErrs, field.Required(specPath.Child("storageClassName"), "LocalStorage volumes require a StorageClass"))
+			}
+			// hostname label is required
+			hostname, found := pv.Labels[metav1.LabelHostname]
```
+1 for admission controller.

We should trust the admin that correct labels are set or the admission plugin is running. To be extra careful, SetUp() call of  volume plugin for LocalStorage could check that it runs on the right node before mount.

jsafrane commented:
LGTM for types.go and validation

jsafrane commented on pkg/api/types.go:
```diff
@@ -370,6 +370,9 @@ type PersistentVolumeSource struct {
 	ScaleIO *ScaleIOVolumeSource
+	// LocalStorage represents a local volume accessible only from one node
+	// +optional
+	LocalStorage *LocalStorageVolumeSource
```
`LocalStorage` -> `Local`? Or `LocalVolume`, following `PortworxVolume` and `PortworxVolumeSource` precedent?

msau42 replied:
I could go with `LocalVolume`.

msau42 commented:
Hmm. I just realized that hostname is not necessarily the same as node name.  So the approach of using the hostname label will not work because the scheduler operates on node name, of which there is no label.

We could add a node name label to the node object, but that seems kind of redundant.  It would fit into the current design of using topologyKey though.

msau42 commented on pkg/api/validation/validation.go:
```diff
@@ -1067,7 +1086,8 @@ var supportedAccessModes = sets.NewString(string(api.ReadWriteOnce), string(api.
 
 func ValidatePersistentVolume(pv *api.PersistentVolume) field.ErrorList {
-	allErrs := ValidateObjectMeta(&pv.ObjectMeta, false, ValidatePersistentVolumeName, field.NewPath("metadata"))
+	metaPath := field.NewPath("metadata")
```
this can be reverted since we're not checking labels anymore

msau42 commented on pkg/apis/extensions/types.go:
```diff
@@ -918,6 +918,7 @@ var (
 	Projected             FSType = "projected"
 	PortworxVolume        FSType = "portworxVolume"
 	ScaleIO               FSType = "scaleIO"
+	LocalStorage          FSType = "localStorage"
```
This file is for podsecuritypolicy, where you can control which pods are allowed to use what types of volumes.  However, it only seems to work for direct volumes, and not persistent volumes.  So I'm not sure if this is needed, unless there are plans to also enforce persistent volume types.  @timothysc

vishh replied:
+1 for removing it for now. Access can also be restricted using quota for PVs

msau42 commented on pkg/security/podsecuritypolicy/util/util.go:
```diff
@@ -65,6 +65,7 @@ func GetAllFSTypesAsSet() sets.String {
 		string(extensions.Projected),
 		string(extensions.PortworxVolume),
 		string(extensions.ScaleIO),
+		string(extensions.LocalStorage),
```
Same thing, may not be needed.

vishh commented on pkg/api/types.go:
```diff
@@ -385,6 +388,9 @@ const (
 	BetaStorageClassAnnotation = "volume.beta.kubernetes.io/storage-class"
+	// AlphaStorageTopologyAnnotation is set in the StorageClass and gives the scheduler
+	// a hint that the volumes in this StorageClass have a topology constraint
+	AlphaStorageTopologyAnnotation = "storage.alpha.kubernetes.io/topology-key"
```
What is the value for this key? We can only recognize "zones" and "nodes" as of now across the system. We need to mention what the possible values are.

msau42 replied:
The value can be any label that exists on the PV + Node.  I can add that to the comment.  

The pod affinity feature specifies label selector and topologyKey as separate fields.  Maybe we can do the same here.
{{LabelSelector: P1, TopologyKey: "node"}}

vishh replied:
As discussed offline, I think we can use labels for expressing topology. Here is the algorithm that we discussed:

1. PVs expose a single topology label - `topology.kubernetes.io/node:foo`, or `topology.k8s.io/zone:bar`
2. PVCs can already select using this topology label. Storage Classes should include a Label Selector that can specify topology constaints.
```yaml
kind: StorageClass
spec:
  selector:
    - key: "topology.k8s.io/node"
```
or 
```yaml
kind: StorageClass
spec:
  selector:
    - key: "topology.k8s.io/zone"
      operator: In
      values:
      - bar-zone
```
3. Nodes will expose all aspects ot topology via consistent label keys.
```yaml
kind: Node
metadata:
  label: 
   topology.kubernetes.io/node : foo
   topology.k8s.io/zone : bar
```
4. The scheduler can then combine NodeSelector on the pod with the Selector on the StorageClass while identifying Nodes that can meet the storage locality requirements. 

This method would require using consistent label keys across nodes and PVs. I hope that's not a non starter. 

cc @thockin Does this align with your thoughts on topology?

msau42 replied:
I'm not sure about the behavior of just specifying the key without an operator or value.  For pod affinity, they actually have two fields, a label selector and a topologyKey.  The topologyKey field means to match the labels between two objects (in this case, PV and Node).  Whereas the selector is applying the filters to a single object.

I also spoke with @lavalamp about using labels to express the topology constraints.  Would we ever want to express in the PV more than a single label/value?  Like, this PV can be accessed on node1,2, and 3?  Or define a new single label "mylabel=foo", and put that same label/value on node1,2,3?

vishh replied:
Sorry, may be label selector is not the right keyword. We definitely need a better selection scheme similar to that of "Affinity/Anti-affinity". I think we need value in the case of storage in addition to just a key. 

> Like, this PV can be accessed on node1,2, and 3

In such a scenario, a PV can include multiple "node" topology labels.

msau42 replied:
I'm going to remove this annotation from this PR for now because there is still more discussion on this topic that we have to do, and I don't want it to block everything else.  Let's use #43504 to continue the discussion.

vishh commented on pkg/api/types.go:
```diff
@@ -1202,6 +1208,21 @@ type KeyToPath struct {
+	NodeName string
+	// Fs represents a filesystem-based local volume
+	// +optional
+	Fs *LocalFsVolume
```
Did you consider having the "access types" (block or fs) as a separate field instead of this approach? 

```go
type LocalVolumeSource struct {
 NodeName string
 Path string
 AccessType string // possible values are "block" or "fs"
}
```
I see `AccessType` becoming an unique field in Storage Class and PVC as well and so we can have a consistent API

msau42 replied:
@jsafrane prefers not to overload the Path field.  It can also make sense to have two separate types if they end up requiring different fields to be specified.

If for block support, we end up adding some access type field, I think it would make sense for that to be in the PV and not the specific volume type.

vishh replied:
Why is path to a block device file not ideal?
Isn't the accessType tied to the volume? A PV cannot expose a block device as a filesystem without formatting it. So wouldn't a PV still want to know what the access type is for an underlying volume?

msau42 replied:
Path to a block device makes sense, but we're not sure yet if there may be any other additional fields required.  So we prefer to have two different types to make field expansion easier in the future.

We are looking at exposing PVs as block devices in the future, so I think having that type field in the PV would make more sense.

vishh replied:
> but we're not sure yet if there may be any other additional fields required

That's ok. Path will be forwards compatible right?

> We are looking at exposing PVs as block devices in the future

If the assumption is that PV is representing an attribute of the underlying storage type, then that's OK. In theory, a storage type should specify it's access type. A PV is merely a wrapper around an existing storage type.

msau42 replied:
I mainly want to avoid a situation like:  Fields A, C, and F are for file-based volumes.  Fields A, B, and D are for block-based volumes.

vishh replied:
Is this a hypothetical scenario or do you have concrete examples?

msau42 replied:
It's hypothetical, as we don't have a detailed investigation or plan for block support.  But I think it is a cleaner API to clearly separate the two types of sources, and also makes future expansion easier.

thockin replied:
Can we sync with Saad on CSI - overloading path for block devs is exactly what he's planning, i think.

msau42 replied:
@saad-ali can you weigh in here? If you are also planning to overload the path for block and file, then I should do the same here to be consistent.

jsafrane replied:
> Is this a hypothetical scenario or do you have concrete examples?

Local volume proposal wants to cover both devices mounted outside of Kubernetes and block devices mounted by Kubernetes (or presented as raw devices to pods somehow). To me, these are two quite different volumes. The first one needs a path to local directory, the second one needs path to a device name (which is not a directory) and at least a `fsType` that should be created there if it's going to be used as regular PV (i.e. mounted by Kubernetes). Hence separate structures `FS` and `Block` (in future).

Overloading `Path` with both mount directory and device path looks really ugly to me.

vishh commented on pkg/api/validation/validation.go:
```diff
@@ -1237,6 +1256,21 @@ func ValidatePersistentVolume(pv *api.PersistentVolume) field.ErrorList {
+			}
+			allErrs = append(allErrs, validateLocalVolumeSource(pv.Spec.LocalVolume, specPath.Child("localVolume"))...)
+			// StorageClass is required
+			if len(pv.Spec.StorageClassName) == 0 {
```
Why is storage class required?

msau42 replied:
Without a storageclass, the scheduler won't know to look for a storage topology and the pod will get scheduled to any node.

vishh replied:
How will storage topology be reflected in PVCs? I feel we need to include support for PVCs too without storage class. Am I the only one thinking if this is necessary?

msau42 replied:
The main concern I have is that PVC Is a user construct, and the user shouldn't care about the underlying storage topology.  We could push towards a model where StorageClass is required in order to express more advanced storage concepts.

ddysher replied:
Trying to understand why storageclass is needed here; say we have a PV created by our 'secondary partition discovery daemon':

```yaml
kind: PersistentVolume
apiVersion: v1
metadata:
  name: local-pv-1
  labels:
    kubernetes.io/hostname: node-1
spec:
  capacity:
    storage: 100Gi
  localStorage:
    fs:
      path: /var/lib/kubelet/storage-partitions/local-pv-1
  accessModes:
    - ReadWriteOnce
  persistentVolumeReclaimPolicy: Delete
  storageClassName: local-fast
```

Following is the storageclass we have:
```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: local-fast
provisioner: ""
toplogyKey: kubernetes.io/hostname
```

Now user creates a PVC to claim a local volume:

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: myclaim
spec:
  storageClaasName: local-fast
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 8Gi
```

PV/PVC controller will not bind the PVC; rather, when user creates a Pod:

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: mypod
spec:
  containers:
    - name: localvolume-pod
      image: nginx
      volumeMounts:
      - mountPath: "/var/www/html"
        name: mypd
  volumes:
    - name: mypd
      persistentVolumeClaim:
        claimName: myclaim
```

At this moment, scheduler will get the PVC `myclaim`, find suitable PVs, then make a desicision based on topologyKey and other constraints like cpu request, then bound the PV and PVC.

Just started reading the proposal today and haven't yet understand the full workflow yet, but a couple of question off my head,

- How PV/PVC controller knows NOT to bind the PVC? There should be a way to differentiate `local` and `remote`. I believe this is discussed in the proposal, but I don't see any API defined here, do we need another field?
- PV is created by discovery daemon, how does it know what storageClassName for PVs it created?
- PV and PVC both have storageClassName, is this required? am i missing something?

vishh replied:
Should we "Recommend" instead of "push"?

msau42 replied:
I can leave out the verification for now until we have a better idea of how the scheduling is going to work.  But I don't think the PVC (user) should be aware of topology constraints of the underlying storage.

ddysher replied:
Trying to understand why storageclass is needed here; say we have a PV created by our 'secondary partition discovery daemon':

```yaml
kind: PersistentVolume
apiVersion: v1
metadata:
  name: local-pv-1
  labels:
    kubernetes.io/hostname: node-1
spec:
  capacity:
    storage: 100Gi
  localStorage:
    fs:
      path: /var/lib/kubelet/storage-partitions/local-pv-1
  accessModes:
    - ReadWriteOnce
  persistentVolumeReclaimPolicy: Delete
  storageClassName: local-fast
```

Following is the storageclass we have:
```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: local-fast
provisioner: ""
toplogyKey: kubernetes.io/hostname
```

Now user creates a PVC to claim a local volume:

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: myclaim
spec:
  storageClaasName: local-fast
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 8Gi
```

PV/PVC controller will not bind the PVC; rather, when user creates a Pod:

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: mypod
spec:
  containers:
    - name: localvolume-pod
      image: nginx
      volumeMounts:
      - mountPath: "/var/www/html"
        name: mypd
  volumes:
    - name: mypd
      persistentVolumeClaim:
        claimName: myclaim
```

At this moment, scheduler will get the PVC `myclaim`, find suitable PVs, then make a desicision based on topologyKey and other constraints like cpu request, then bound the PV and PVC.  Not sure if I understand it correctly, but couple of question off my head regarding this workflow:

- How PV/PVC controller knows NOT to bind the PVC? There should be a way to differentiate `local` and `remote`. I believe this is discussed in the proposal, but I don't see any API defined here, do we need another field for the purpose?
- PV is created by discovery daemon, how does it know what storageClassName for PVs it created, do we need admin to configure it?
- PV and PVC both have storageClassName, is this required? am i missing something?

msau42 replied:
* The PV/PVC controller will not bind PVC's that request a storageclass with a topologyKey specified.  This is why if we go with this approach, I think storageclass is required.  Otherwise, we do not know when to trigger this different logic for local volumes.  The alternative is, like you mentioned, another field in the PVC.  But I do not like this alternative, because I think that this is a detail that the user should not need to know.
* The discovery daemon has to be configured to create PVs with a StorageClass name.  I am thinking of having some sort of configmap, with directory -> storageclass mapping.  ie all the volumes under dir1 will be part of storageclass1, everything under dir2 will use storageclass2, etc.
* I think it is required in order to delay the PVC binding until a pod comes.  Maybe for alpha, we could survive without this behavior if you start all your local storage pods first.

ddysher replied:
Thanks for the detailed explanation!

thockin replied:
I'm not sure I follow - is storageclass being used as a flag to change behavior of binding?  Can you explain it for dummies like me? :)

ddysher replied:
And one more question, if it is used as a flag to change behavior of binding, what will be the trigger for unbinding?  Say we have a Pod requesting a PVC, scheduler binds it to a PV.  If we delete the Pod, I don't know if there is a path to trigger an unbind. If not, that means later Pod reusing the PVC, it will be scheduled to the specific Node. But for user, this PVC should be no difference from a new PVC created from scratch.

I think we can either do this statically, i.e. in pv controller, unbind PV/PVC if no Pod is using the PVC; or dynamically, rebind PV/PVC at schedule time. Is this discussed somewhere?

msau42 replied:
@thockin yes, right now the idea is if the StorageClass has a topologyKey specified, then we will trigger this "delayed" binding behavior and also trigger the scheduler predicate to filter nodes based on topologyKey.

We could put the topologyKey in the PVC instead, but I prefer the idea of the StorageClass encapsulating the topology constraints so that the user does not need to be aware.

@ddysher we have no plans to change the unbinding flow.  It will operate the same way it does today.  PVC will be unbound when it is deleted.

Re-using the PVC could actually be a valid use case for data gravity.  You could have one pod that populates the PVC with data.  And another pod come along later to consume that PVC.

ddysher replied:
yes, this is a valid use case.  but what I think we are lacking here is a way for user or admin to unbind pv and pvc such that they can be put back into the scheduling pool. otherwise, we ended up exhausting all scheduling capabilities?

msau42 replied:
So the deleting flow is as follows:

1. User deletes pod
2. User deletes PVC
3. If PV reclaim mode was Retain, then the state becomes "Retain", and the user has to cleanup the PV themselves, delete it, and recreate it.
4. If PV reclaim mode was Delete, then the PV goes into Released state
5. The local storage DaemonSet will cleanup Released PVs and then create new PVs again.

ddysher replied:
:ok_hand: thx for the quick turnaround.

vishh replied:
> We could put the topologyKey in the PVC instead, but I prefer the idea of the StorageClass encapsulating the topology constraints so that the user does not need to be aware.

A user can still manually bind local PVs to PVCs without storage classes. Why force every local PV to expose a storage class?
I'm personally not a fan of tying storage classes to PVs during static provisioning.
In the future, will it be possible to keep these operations separate?
1. Provisioning local volumes
2. Defining storage classes for grouping local volumes

msau42 replied:
I am ok with not requiring StorageClass for the case where the user manually prebinds PVCs to PVs.  What I do not want is the user to have to specify the topologyKey in the PVC instead of StorageClass to trigger the new scheduling/binding flow.

vishh commented on staging/src/k8s.io/client-go/pkg/api/v1/generated.proto:
```diff
@@ -1413,6 +1413,12 @@ message LoadBalancerStatus {
   repeated LoadBalancerIngress ingress = 1;
 }
 
+// LocalFsVolume represents a filesystem-based local volume
```
Are the staging artifacts expected to be committed? cc @caesarxuchao

caesarxuchao replied:
Yep. This file is copied from k8s.io/kubernetes/pkg/api/v1/generated.proto. If they are out of sync, verification-all.sh will fail.

msau42 commented:
@k8s-bot unit test this

ddysher commented on pkg/api/v1/types.go:
```diff
@@ -428,6 +428,9 @@ type PersistentVolumeSource struct {
 	ScaleIO *ScaleIOVolumeSource `json:"scaleIO,omitempty" protobuf:"bytes,19,opt,name=scaleIO"`
+	// LocalVolume represents a local volume accessible only from one node
+	// +optional
+	LocalVolume *LocalVolumeSource `json:"localVolume,omitempty" protobuf:"bytes,20,opt,name=localVolume"`
```
Maybe update the proposal as well?  It uses `LocalStorage`.

msau42 replied:
Yeah the proposal will need some updates after addressing the review comments on this PR.

msau42 commented on pkg/api/types.go:
```diff
@@ -1205,6 +1208,21 @@ type KeyToPath struct {
+// LocalVolume represents a local volume accessible only from one node
+type LocalVolumeSource struct {
+	// NodeName represents the node that the volume is located at
+	NodeName string
```
The thought I had was to have an admission controller add the label to PVs based off of the NodeName field, if it wasn't already set.  There's been some push back against having the system validate specific labels, because theoretically any label could be used, not just the hostname label.

I also want the NodeName to be a required field in the volume spec so I can ensure that the volume is only ever mounted on the correct node, in case some user or system error happens and the volume gets scheduled to the wrong node.

vishh replied:
A local volume is tied to a specific node. So the node is an explicit attribute of that volume and so exposing that as a label isn't a good idea.

msau42 replied:
The only reason for a label is to work with the scheduler's storage topologyKey mechanism, which operates on labels.

ddysher replied:
Since they are the same (label is populated based on NodeName field), why can't we use NodeName directly?  I haven't looked at scheduler part, but to answer my own question, I guess this is to confirm to the same logic applied to affinity/anti-affinity, and can also benefit from code reuse?

msau42 replied:
That's correct, we could use the NodeName directly, and we can do that for now to get something working quickly.   But if we want to do something more generic, and have it work for any arbitrary topology like zones, racks, etc, then the current idea is to use the label approach, similar to how pod affinity/anti-affinity works.

thockin replied:
I'm on the fence.  Local is local to SOMETHING.  Node is arbitrary.  There exist multi-attached hardware devices that can be local to more than one machine (but one at a time).

What happens when we encounter such a beast - is LocalVolume suddenly obsolete?

I don't want to force this idea, since we don't have generalized topo elsewhere yet, but I do want to make sure we think it through properly.  What are concrete concerns?

msau42 replied:
I just want the topology constraint to be required in the volume spec.  This is for security purposes, so if the scheduler makes some mistake or the user forgets to put the topology constraint in the PV, then we are not reverting back to hostpath behavior of mounting arbitrary directories on any node.

So I just want some way to be able to check at mount time, that the volume I am mounting is in fact owned by this node.

I think one way that we can generalize this, is instead of a nodeName field, have a topologyConstraints field that can look something like:

```
TopologyConstraints []TopologyConstraint

type TopologyConstrating struct {
    key string
    value string
}
```

The key-value would essentially be the allowed labels, like "hostname = node1", etc.

In the future, for some storage that could be local to 3 machines, as an example, then you would have 3 entries in this list.

WDYT?

msau42 replied:
Actually I think this would be better as an annotation in the PV, then we can use it for other plugins too like gcepd.  

For local volumes I would still want to require that the topology annotation is set (specific key values are not validated), so that we can ensure that we're mounting the paths on the correct node.

vishh replied:
The constraints have to include a list of values. Example use case - a volume that can be accessed from multiple nodes or racks.

```go
TopologyConstraints []TopologyConstraint

type TopologyConstrating struct {
    key string
    value []string // `or` is the assumed operator
}
```

thockin commented on pkg/api/types.go:
```diff
@@ -370,6 +370,9 @@ type PersistentVolumeSource struct {
 	ScaleIO *ScaleIOVolumeSource
+	// LocalVolume represents a local volume accessible only from one node
+	// +optional
+	LocalVolume *LocalVolumeSource
```
Why have the word "volume" in the name?  Is that intentional or just wasn't considered?  Why not `Local` ?  This is a question, not a request for change (yet)

msau42 replied:
I thought that just "Local" would be too generic.  But I don't feel strongly about either naming option, so am totally open changing it.

vishh replied:
+1 for `Local`.

thockin commented on pkg/api/types.go:
```diff
@@ -1205,6 +1208,21 @@ type KeyToPath struct {
+	NodeName string
+	// Fs represents a filesystem-based local volume
+	// +optional
+	Fs *LocalFsVolume
```
FS

thockin commented on pkg/api/types.go:
```diff
@@ -1205,6 +1208,21 @@ type KeyToPath struct {
+}
+
+// LocalFsVolume represents a filesystem-based local volume
+type LocalFsVolume struct {
```
The idea here is to leave room for block devices, yeah?  What about all the other volume types that didn't do this - can we find a way to solve it for everyone instead of being new for this one?

msau42 replied:
Yeah, consistency with the rest of the volume plugins is important too, so when we do eventually add block support, then we can handle all the volume plugins the same way.  @jsafrane, wdyt?

vishh replied:
> can we find a way to solve it for everyone instead of being new for this one?

Depends on at what level `block` vs `file` is defined. Ideally it has to be surfaced by the underlying storage volume and so this will require API changes to all existing volume types.

thockin commented on pkg/api/v1/types.go:
```diff
@@ -1292,6 +1295,21 @@ type KeyToPath struct {
+	// NodeName represents the node that the volume is located at
+	NodeName string `json:"nodeName" protobuf:"bytes,1,opt,name=nodeName"`
+	// Fs represents a filesystem-based local volume
+	// +optional
```
this says optional but is validates as a one-of.  @mbohlool do we have a preferred way to decorate "one-of" groups?

vishh replied:
I see this in a lot of storage APIs. It makes the APIs hard to interpret.

msau42 replied:
If we go with the approach of flattening the struct and just including Path in the top-level, then this is not an issue anymore.

msau42 commented:
I made the following changes:
- Rename volume type to "Local"
- Get rid of Fs, and just have Path
- Get rid of NodeName, and replace with TopologyConstraints PV annotation.  At least one constraint must be specified for local volumes.
- Remove StorageClass requirement

@vishh @thockin PTAL

vishh commented on pkg/api/helper/helpers.go:
```diff
@@ -594,3 +594,27 @@ func PersistentVolumeClaimHasClass(claim *api.PersistentVolumeClaim) bool {
+
+// GetTopologyConstraintsFromAnnotations gets the json serialized topology constraints data from PersistentVolume.Annotations
+// and converts it to the []TopologyConstraint type in api.
+// TODO: remove when alpha support for topology constraints is removed
```
nit: remove when topology constraints is graduated to beta.

msau42 replied:
Even if we move to beta, we may still have to support alpha for a while?

vishh replied:
Do we really have to? I'd rather have users not use this API if they need forwards compat.

msau42 replied:
We don't have to.  I'm just wondering based off of storageclasses, where we are still "supporting" the alpha annotation even though it's now GA.

vishh replied:
Storage classes are user facing. Topology constraints is not expected to be end user (not cluster admins) facing. So I feel we can drop alpha annotations as soon as we move to beta. Anyways, it was a `nit` :)

msau42 replied:
For beta, we will still have annotations, so I updated the comment to say to update the function when we move to beta.

vishh commented on pkg/api/types.go:
```diff
@@ -388,8 +391,21 @@ const (
+// A label key can have multiple values
+type TopologyConstraint struct {
+	// Key of the label
+	Key string
```
Why are the json annotations missing? Is this part of the "alpha" process?

msau42 replied:
This is the internal API types.go, which does not have json.  The v1 types.go includes the json.

vishh replied:
Sorry. I thought I had deleted that comment.

vishh commented on pkg/api/v1/types.go:
```diff
@@ -437,8 +440,21 @@ const (
+// TopologyConstraint contains label information that the resource is constrained to
+// A label key can have multiple values
+type TopologyConstraint struct {
+	// Key of the label
```
Should we list the supported topology constraints? I assume that you will update kubelet to expose certain first class keys by default.

msau42 replied:
I thought the idea was the user can use any label that is on the node, and we won't have any first-class keys.

ddysher replied:
Even though the idea is to allow user to use any label, the type is named "Topology" and we do have 'first-class' [keys](https://kubernetes.io/docs/reference/labels-annotations-taints/).  I guess for the most of time, user will just use 'kubernetes.io/hostname'.

vishh replied:
@msau42 in-order for local PVs to work, we need to settle on a special label that k8s will expose by default from kubelets. That's a first class label type right? I'm saying that we need to document that.

msau42 replied:
Can it not use existing node labels, like "kubernetes.io/hostname", or any other label that the user defines?  I don't think we need code to explicitly check specific labels.

vishh replied:
I'm not suggesting adding code for validation. But, the documentation should suggest that there are some special topology keys that are supported by upstream by default.
`Hostname` is not portable right? Don't we need `NodeName`?

vishh replied:
https://kubernetes.io/docs/reference/labels-annotations-taints/

msau42 replied:
There's another debate going on about whether a NodeName label on a Node object makes sense.  The other pod scheduling features all use the hostname label, so if we added a nodename label, that could potentially also add some consistency confusion when using local PVs with pod affinity.  I would rather handle it separately than block this change.

vishh replied:
OK. I'm just a bit weary of saying all local PVs need **some** topology constraints, without being prescriptive about it. It's an alpha API, so as you suggest we can bikeshed on that for now.

msau42 replied:
I understand your concern.  Definitely before we release, we will need to figure out the hostname vs nodename issue and document which one we recommend to use.

vishh commented on pkg/api/v1/types_swagger_doc_generated.go:
```diff
@@ -1147,6 +1156,7 @@ var map_PersistentVolumeSource = map[string]string{
 	"photonPersistentDisk": "PhotonPersistentDisk represents a PhotonController persistent disk attached and mounted on kubelets host machine",
 	"portworxVolume":       "PortworxVolume represents a portworx volume attached and mounted on kubelets host machine",
 	"scaleIO":              "ScaleIO represents a ScaleIO persistent volume attached and mounted on Kubernetes nodes.",
+	"local":                "Local represents a local VFS volume with topology constraints",
```
`Local represents a directory with specific topology constraints`?

vishh commented on pkg/api/validation/validation.go:
```diff
@@ -1238,6 +1250,22 @@ func ValidatePersistentVolume(pv *api.PersistentVolume) field.ErrorList {
+			allErrs = append(allErrs, validateLocalVolumeSource(pv.Spec.Local, specPath.Child("local"))...)
+
+			// TopologyConstraints is required
+			if !topologyExists {
```
I assume this check is necessary to handle dynamic binding in which case it might be useful to define some standard keys too (`topology.storage.kubernetes.io/node` for example)

msau42 replied:
It's a safety net so that before kubelet mounts a volume, we can verify that the volume belongs to the node.  Without it, there is no sense of node "ownership" of a volume, and the system could potentially mount arbitrary volumes on incorrect nodes.

msau42 replied:
I understand your question now.  Yes, it is also used when making the binding decision to filter which PVs are accessible from a particular node.

vishh replied:
As mentioned earlier, we need to define at-least the topology keys that are reserved for local PVs.

vishh reviewed (changes requested):
Just a couple of nits. I like this version. If @thockin is OK with this API, it's ready to be merged.

jsafrane commented on pkg/api/types.go:
```diff
@@ -1205,6 +1221,13 @@ type KeyToPath struct {
+type LocalVolumeSource struct {
+	// The full path to the volume on the node
+	// For alpha, this path must be a file-based volume
+	Path string
```
This is not what has been agreed in the proposal:

> [some] addon DaemonSet can also discover block devices and creates corresponding PVs for them with the `block` field.  
>```
>kind: PersistentVolume
>...
>spec:
>      localStorage:
>         block:
>           device: /var/lib/kubelet/storage-raw-devices/foo
>```

@vishh, @thockin, how are you going to extend LocalVolumeSource with block devices in future? Overloading `Path` with both path to a mounted directory and a device name looks ugly to me, it has totally different semantics. In addition, you need to put at least `fsType` somewhere into the structure and it will look weird for already mounted directories where it does not make any sense.

msau42 replied:
The main argument for overloading is to be consistent with the other volume types.  For example, GCE PD does not distinguish between if the disk has already been formatted or not.  I agree that for block case, we need to add fsType to the structure, similar to how gcepd does it.  I wanted to leave that out for now until we have block support.

jsafrane replied:
GCE PD is **always** a block device and Kubernetes makes a fs on it and mounts it. With LocalVolumeSource, now it's a filesystem (or even a shared directory) formatted and mounted outside of Kubernetes and we should have a clean way how to extend it to block device formatted and mounted by Kubernetes.

msau42 replied:
Had a discussion with @vishh and @saad-ali, here are our thoughts on how to support file and block based volume in a generic manner:

The basic idea is that the PVC/PV contains information about how to consume the volume (as a file or block), and the specific volume spec contains the "connection" information to the underlying storage.

Have a volumeType field in the PersistentVolume, which is set at creation time. 

```
kind: PersistentVolume
spec:
    volumeType: file or block
    local:
       # If we renamed path to sourcePath, does that help at all?
       path: /dir1
```

Then, in the PVC, the user specifies how they want to consume the volume
```
kind: PersistentVolumeClaim
spec:
    volumeType: file or block
    # This is only valid for file-based
    fsType: ext4
```

If the PVC wants file (which is today's behavior), and its bound PV is block type, then the system will create the fs.  If the bound PV is already fs type, then the system doesn't format anything.  This is different from today's model, where the fs is specified in the specific volume spec, but I think by moving the volume type into PV/PVC, that will make it simpler to support block access in the future.  This is also consistent with how CSI plans to support block access.

In the future, if PVC wants block, then it can only be bound to block-based PVs.

Please let me know if this makes sense. We can also schedule a meeting to discuss further if you like.

mtanino replied:
I'm interested in this discussion too.

Above proposal looks good to me but I have a question. 
Is the volumeType parameter mandatory required?
Or user can skip the parameter then the default value is set to "file".

msau42 replied:
For backwards compatibility I don't think we cannot require it.  I don't think it can be defaulted either.  Something like nfs is only file-based, and gce pd is only block-based.  So if it's left out, then we could resort to old behavior that it's up to the volume plugin to do the formatting.

jsafrane replied:
I totally agree that `pv.spec.volumeType` and `pvc.spec.volumeType: file or block` makes sense and is the right way to go (in a separate PR/proposal), but I am commenting something else.

Right now, `LocalVolumeSource` covers only filesystems mounted outside of Kubernetes. To some extent it's very similar to HostPath. `LocalVolumeSource.PathToMountedDirectory` is the only thing we need here.

In future, we want `LocalVolumeSource` to cover also block devices, mounted by Kubernetes. This is similar to GCE PD or AWS EBS, just without attach/detach, there already is a device in /dev. We will need `LocalVolumeSource.PathToDevice` (which is IMO something different than `PathToMountedDirectory`) and `FSType`. Again, overloading `LocalVolumeSource.Path` for both paths is bad, its semantics changes based on what kind of local volume it is. That's why I think two structures (directory / block) are better than one. It's clear what `Path` means and it's clear that fsType applies only to local block volumes.

msau42 replied:
I guess the main issue today is that this is the first volume plugin that can export both a file or block volume.  All other plugins today are only block or only file.  So the options I see are:

1.  Having a file struct and a block struct within the same volume source object.  None of the other volume objects do this, so it looks inconsistent.
2.  Overload Path field and use PV.VolumeType field to distinguish between the two.  The Path field having different behaviors/semantics depending on the type may be difficult for users to understand.
3.  Have two completely separate volume plugins, one for local file volumes, and one for local block volumes.  This is more consistent with the current semantics (kinda like rbd and cephfs).  We are trying not to hardcode the Local volume type into the scheduler and pv controller, so adding a block volume plugin in the future should work as long as it uses the same general topology mechanisms.

What do you guys think about the 3rd approach?

On a side note, I think when we support local block devices, and we go with the new block access model, then we won't need fstype in the volume spec.  We'll just use the fstype in the PVC.

saad-ali replied:
Let's be very crisp about the purpose of the information we are trying to find a home for in the API. Doing so will help make it easier to justify where in the API it makes sense.

The way I see it, we need to differentiate between two pieces of information:
1. Generic volume information
    * Information Kubernetes must understand and operate on in order to complete the volume setup process.
2. Plugin specific information
    * Relevant to individual volume plugins and (likely) different between volume plugins.
    * Opaque to Kubernetes: only plugin needs to understand this.
    * In the future with CSI plugins this will likely no longer be part of the Kubernetes API, and will be demoted to opaque pass through key/value pairs.

In this thread we have the following pieces of information:
1) Volume consumption information: block OR mount
    * Indicates if volume will be consumed as a block device OR as mounted volume.
    * If block device, no further information required about consumption.
    * If mounted volume, must also specify:
      * File system
      * Additional mount flags
    * *This is "Generic volume information" as defined above, because k8s must be aware of this to figure out how to properly surface the volume in the container (block device or mounted directory).*
2) Volume connection information: information required by plugin to identify volume
    * This is "plugin specific information" as defined above, because k8s does not care what it is (only the volume plugin does).
    * For "local storage" volume plugin, this includes  "volume source" information:
      * For block device local storage, this is the "block device path".
      * For mounted volumes local storage, this is the "path to mounted directory".

Based on this breakdown, here are my thoughts on @msau42's options:
> 1. Having a file struct and a block struct within the same volume source object. None of the other volume objects do this, so it looks inconsistent.

Not a good option. It requires the k8s to understand and use volume plugin specific fields to figure out how to surface the volume (these fields should only be used by the volume plugin)

> 2. Overload Path field and use PV.VolumeType field to distinguish between the two. The Path field having different behaviors/semantics depending on the type may be difficult for users to understand.

Ok option. Since the "source/connection information" is inside the plugin specific fields, and the "consumption information (block or mount)" is in the generic k8s volume fields. 

> 3. Have two completely separate volume plugins, one for local file volumes, and one for local block volumes. This is more consistent with the current semantics (kinda like rbd and cephfs). We are trying not to hardcode the Local volume type into the scheduler and pv controller, so adding a block volume plugin in the future should work as long as it uses the same general topology mechanisms.

Ok as long as we leave the mount vs block info in PV/PVC (since it is required by Kubernetes), but doing so AND having separate "block only" and "mount only" local storage plugins is very ugly.

So I perfer @msau42's second option, but I understand @jsafrane's concerns, and would also be ok with his earlier proposal to break "path" into two volume specific fields inside the local storage volume: `LocalVolumeSource.PathToDevice` and `LocalVolumeSource.PathToMountedDirectory` (although that looks a little ugly too).

msau42 replied:
Thanks Saad for the detailed explanation and highlighting the pros and cons of each approach.  I think we won't be able to come up with a scheme that perfectly addresses each concern, so I am leaning towards option 2, or having two explicit fields, PathToDevice, or PathToMountedDirectory.  @jsafrane what do you think?

jsafrane replied:
3) looks quite ugly, so 2 looks the best from the list. Can you add a comment that `Path` is going to be reused both for a dir and device so the one who's going to approve the API changes (bgrant?) is aware of it?

vishh commented:
Can it be assumed that a filesystem can be placed on top of all storage
volumes that provide block level access? And what about dealing with a list
of specific filesystems that are recommended for different block devices?

On Thu, May 11, 2017 at 12:48 PM, Michelle Au <notifications@github.com>
wrote:

> *@msau42* commented on this pull request.
> ------------------------------
>
> In pkg/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/44640#discussion_r116085321>
> :
>
> > @@ -1205,6 +1221,13 @@ type KeyToPath struct {
>  	Mode *int32
>  }
>
> +// Local represents a directory with specific topology contraints
> +type LocalVolumeSource struct {
> +	// The full path to the volume on the node
> +	// For alpha, this path must be a file-based volume
> +	Path string
>
> Had a discussion with @vishh <https://github.com/vishh> and @saad-ali
> <https://github.com/saad-ali>, here are our thoughts on how to support
> file and block based volume in a generic manner:
>
> The basic idea is that the PVC/PV contains information about how to
> consume the volume (as a file or block), and the specific volume spec
> contains the "connection" information to the underlying storage.
>
> Have a volumeType field in the PersistentVolume, which is set at creation
> time.
>
> kind: PersistentVolume
> spec:
>     volumeType: file or block
>     local:
>        # If we renamed path to sourcePath, does that help at all?
>        path: /dir1
>
> Then, in the PVC, the user specifies how they want to consume the volume
>
> kind: PersistentVolumeClaim
> spec:
>     volumeType: file or block
>     # This is only valid for file-based
>     fsType: ext4
>
> If the PVC wants file (which is today's behavior), and its bound PV is
> block type, then the system will create the fs. If the bound PV is already
> fs type, then the system doesn't format anything. This is different from
> today's model, where the fs is specified in the specific volume spec, but I
> think by moving the volume type into PV/PVC, that will make it simpler to
> support block access in the future. This is also consistent with how CSI
> plans to support block access.
>
> In the future, if PVC wants block, then it can only be bound to
> block-based PVs.
>
> Please let me know if this makes sense. We can also schedule a meeting to
> discuss further if you like.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/44640#discussion_r116085321>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKIFpMrkLtkrSnyD7PEvL01w2aPJ2ks5r42X5gaJpZM4NA_gk>
> .
>

msau42 commented:
The only incompatibility issue I can think of is fstrim support.  That requires kernel and device support.  But that doesn't impact basic filesystem functionality.  It just means the fstrim command won't work, which I think is ok.  If anyone else knows about any incompatibility issues, please let us know.

Preferred filesystems could potentially be included in the storageclass.  I don't think it belongs in the PVC or PV.

msau42 commented:
@thockin @vishh I added topology constraints to the PV object and added a comment saying that path will be used for both file and block devices.  PTAL

msau42 commented:
@k8s-bot kops aws e2e test this

vishh commented on pkg/api/types.go:
```diff
@@ -388,8 +391,21 @@ const (
+	AlphaStorageTopologyConstraintsAnnotation = "volume.alpha.kubernetes.io/topology-constraints"
 )
 
+// TopologyConstraint contains label information that the resource is constrained to
```
```
TopologyConstraint represents topology dimensions for a resource. It consists of a `Topology key` that represents a unique topology dimension and one or more `optional values` that further constrain a specific topology dimension.
Example: Topology key `kubernetes.io/zone` can have values `foo` & `bar` where the resource that this topologyconstraint applies to is accessible only from zones foo and bar.
``` 

This is no longer a traditional `label`. Maybe say `TopologyConstaints are expected to be similar to labels. They are not managed by the system`?

msau42 replied:
Regarding 'optional values'.  I was assuming at least one value must be specified.  What is the behavior if no values are specified?

msau42 replied:
Should we mention that the topology key/values are matched up with corresponding labels on the Node?

vishh reviewed (changes requested):
A couple more comments. This is ready to be merged once these comments are addressed.

vishh commented:
This feature needs a dedicated doc page and we should cross link to that
from here.

On Tue, May 16, 2017 at 2:02 PM, Michelle Au <notifications@github.com>
wrote:

> *@msau42* commented on this pull request.
> ------------------------------
>
> In pkg/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/44640#discussion_r116856865>
> :
>
> >  )
>
> +// TopologyConstraint contains label information that the resource is constrained to
>
> Should we mention that the topology key/values are matched up with
> corresponding labels on the Node?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/44640#discussion_r116856865>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKELswednIOnqkPoXv1BulMp7F8_rks5r6g7ugaJpZM4NA_gk>
> .
>

msau42 commented:
We do have a documentation task on our issue tracker, so it will be done before releasing.

msau42 commented:
@k8s-bot kubemark e2e test this

vishh commented:
/lgtm

vishh commented:
@thockin this PR is ready to go once you are happy with it.

msau42 commented:
/release-note-none

msau42 commented:
@k8s-bot kops aws e2e test this

msau42 commented on pkg/api/types.go:
```diff
@@ -388,8 +391,24 @@ const (
 	MountOptionAnnotation = "volume.beta.kubernetes.io/mount-options"
+
+	// StorageTopologyConstraint defines topology constraints of a PV
+	// Value is type []TopologyConstraint
```
I'm thinking of replacing this with the NodeAffinity type instead, since I think this can be expressed through NodeSelector.  Then we don't have to add this new type for topology constraints.  For alpha at least and maybe even beta, we will only implement the RequiredDuringSchedulingIgnoredDuringExecution NodeSelector part and ignore/error on the other fields.

What do you guys think?

vishh commented:
+1 for using NodeAffinity with additional documentation and validation. In
any case, I'd stick to annotations for v1.7.

On Thu, May 18, 2017 at 10:29 AM, Michelle Au <notifications@github.com>
wrote:

> *@msau42* commented on this pull request.
> ------------------------------
>
> In pkg/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/44640#discussion_r117306885>
> :
>
> > @@ -388,8 +391,24 @@ const (
>
>  	// MountOptionAnnotation defines mount option annotation used in PVs
>  	MountOptionAnnotation = "volume.beta.kubernetes.io/mount-options"
> +
> +	// StorageTopologyConstraint defines topology constraints of a PV
> +	// Value is type []TopologyConstraint
>
> I'm thinking of replacing this with the NodeAffinity type instead, since I
> think this can be expressed through NodeSelector. Then we don't have to add
> this new type for topology constraints. For alpha at least and maybe even
> beta, we will only implement the RequiredDuringSchedulingIgnoredDuringExecution
> NodeSelector part and ignore/error on the other fields.
>
> What do you guys think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/44640#pullrequestreview-39004238>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKAZaWLkFPyzAWrRV55YQkBmVyV4cks5r7H_dgaJpZM4NA_gk>
> .
>

msau42 commented:
I updated to use NodeAffinity in the annotation.  @vishh @thockin PTAL

vishh commented:
/lgtm

thockin commented:
/approve

msau42 committed:
Generated code for local persistent volumes

msau42 commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

thockin commented:
/lgtm
